crates/uv-resolver/src/lock.rs

```diff
@@ -0,0 +1,753 @@
+        }
+    }
+
+    fn from_built_dist(built_dist: &BuiltDist) -> Source {
+        match *built_dist {
+            BuiltDist::Registry(ref reg_dist) => Source::from_registry_built_dist(reg_dist),
+            BuiltDist::DirectUrl(ref direct_dist) => Source::from_direct_built_dist(direct_dist),
+            BuiltDist::Path(ref path_dist) => Source::from_path_built_dist(path_dist),
+        }
+    }
+
+    fn from_source_dist(source_dist: &distribution_types::SourceDist) -> Source {
+        match *source_dist {
+            distribution_types::SourceDist::Registry(ref reg_dist) => {
+                Source::from_registry_source_dist(reg_dist)
+            }
+            distribution_types::SourceDist::DirectUrl(ref direct_dist) => {
+                Source::from_direct_source_dist(direct_dist)
+            }
+            distribution_types::SourceDist::Git(ref git_dist) => Source::from_git_dist(git_dist),
+            distribution_types::SourceDist::Path(ref path_dist) => {
+                Source::from_path_source_dist(path_dist)
+            }
+        }
+    }
+
+    fn from_registry_built_dist(reg_dist: &RegistryBuiltDist) -> Source {
+        Source::from_index_url(&reg_dist.index)
+    }
+
+    fn from_registry_source_dist(reg_dist: &RegistrySourceDist) -> Source {
+        Source::from_index_url(&reg_dist.index)
+    }
+
+    fn from_direct_built_dist(direct_dist: &DirectUrlBuiltDist) -> Source {
+        Source {
+            kind: SourceKind::Direct,
+            url: direct_dist.url.to_url(),
+        }
+    }
+
+    fn from_direct_source_dist(direct_dist: &DirectUrlSourceDist) -> Source {
+        Source {
+            kind: SourceKind::Direct,
+            url: direct_dist.url.to_url(),
+        }
+    }
+
+    fn from_path_built_dist(path_dist: &PathBuiltDist) -> Source {
+        Source {
+            kind: SourceKind::Path,
+            url: path_dist.url.to_url(),
+        }
+    }
+
+    fn from_path_source_dist(path_dist: &PathSourceDist) -> Source {
+        Source {
+            kind: SourceKind::Path,
+            url: path_dist.url.to_url(),
+        }
+    }
+
+    fn from_index_url(index_url: &IndexUrl) -> Source {
+        match *index_url {
+            IndexUrl::Pypi(ref verbatim_url) => Source {
+                kind: SourceKind::Registry,
+                url: verbatim_url.to_url(),
+            },
+            IndexUrl::Url(ref verbatim_url) => Source {
+                kind: SourceKind::Registry,
+                url: verbatim_url.to_url(),
+            },
+            IndexUrl::Path(ref verbatim_url) => Source {
+                kind: SourceKind::Path,
+                url: verbatim_url.to_url(),
+            },
+        }
+    }
+
+    fn from_git_dist(git_dist: &GitSourceDist) -> Source {
+        // FIXME: Fill in the git revision details here. They aren't in
+        // `GitSourceDist`, so this will likely need some refactoring.
```

Comment:
I also ran into this problem a bunch; we're tracking some information on the urls of the dist types but it's not readily accessible